### PR TITLE
MAGN-5931: Insert pasted elements at the current workspace offset

### PR DESCRIPTION
**Pull request by pboyer**
### Issue

Currently, when users paste content into a workspace, the original position of the pasted content determines the new location.  

This is problematic for several reasons:

1.  It results in the new content being pasted, confusingly, directly atop the old content.  
2.  It doesn't reflect the current workspace offset.  As a result, a user that has panned significantly will have their pasted content offscreen.  
3.  Two workspaces seldomly have the same offset.  This causes 2 to occur when copying between workspaces.

### Fix 

Use the current workspace offset in order to determine the location of the new content.  In this pull request, the top left corner of the copied content is determined and all of the pasted content is shifted so as to consistently place the pasted content in view.  

Further, a small offset is introduced between subsequent paste operations in order to prevent issue 1 when  copying repeatedly.

### Reviewer

- [ ] @ikeough

### CODE REVIEW

File: src/DynamoCore/Models/DynamoModel.cs [hunk 1]
```diff
@@ -1061,13 +1077,50 @@ public void Paste()
...
+                newNodeModels.Add( newNode );
+
+                minX = Math.Min(node.X, minX);
+                minY = Math.Min(node.Y, minY);
+            }
+
+            // Move all of the notes and nodes such that they are aligned with
+            // the top left of the workspace
+            var workspaceX = -CurrentWorkspace.X / CurrentWorkspace.Zoom;
+            var workspaceY = -CurrentWorkspace.Y / CurrentWorkspace.Zoom;
```
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: This computes the current workspace offset in true workspace coordinates.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 2]
```diff
@@ -1061,13 +1077,50 @@ public void Paste()
...
+            CurrentWorkspace.IncrementPasteOffset();
+
+            var shiftX = workspaceX - minX + CurrentWorkspace.CurrentPasteOffset;
+            var shiftY = workspaceY - minY + CurrentWorkspace.CurrentPasteOffset;
+
+            foreach (var model in newNodeModels.Concat<ModelBase>(newNoteModels))
+            {
+                model.X = model.X + shiftX;
+                model.Y = model.Y + shiftY;
+            }
```
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: Here we shift the new content before it is actually inserted into the workspace.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 3]
```diff
@@ -1061,13 +1077,50 @@ public void Paste()
...
             OnRequestLayoutUpdate(this, EventArgs.Empty);
 
+            // Add the new NoteModel's to the Workspace
+            foreach (var newNote in newNoteModels)
+            {
+                CurrentWorkspace.AddNote(newNote, false);
+                createdModels.Add(newNote);
+                AddToSelection(newNote);
+            }
+
```
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: Now we actually insert the new content into the workspace.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 4]
```diff
@@ -1091,29 +1144,6 @@ public void Paste()
...
-            foreach (var newNote in newNotes)
-            {
-                CurrentWorkspace.AddNote(newNote, false);
-                createdModels.Add(newNote);
-                AddToSelection(newNote);
-            }
-
-            foreach (var de in nodeLookup.Values)
-                AddToSelection(de);
-
```
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: This code has been moved above and simplified.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 5]
```diff
@@ -1022,6 +1022,10 @@ public void Copy()
         /// </summary>
         public void Paste()
         {
+            //clear the selection so we can put the
+            //paste contents in
+            DynamoSelection.Instance.ClearSelection();
+
```
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: I think this state change should be front & center, rather than buried in local variable definitions.

**Comment by ikeough** [comment]
@pboyer This is nice. LGTM.

**Comment by pboyer** [comment]
FYI @jnealb & @riteshchandawar.

**Pull request merged**